Feature/workflow tracking service connector

KochTobi opened the pull request:
## What is this PR about?
This Pull Request introduces an implementation for the WorkflowTrackingDataSource Interface.
The implemented WorkflowTrackingServiceConnector connects to the nextflow-logger-service.
## Changes made
* modify WorkflowTrackingDataSource interface
* implement WorkflowTrackingServiceConnector
* add constructor for Service object to WorkflowTrackingService helper class
* fix pom.xml to run with spock framework

KochTobi committed:
WorkflowTrackingDataSource implementation for nextflow-logger-service

modify WorkflowTrackingDataSource interface
implement WorkflowTrackingServiceConnector
add constructor for Service object to WorkflowTrackingService helper class

KochTobi committed:
Variable name change

change url_string to urlString

sven1103 commented on src/main/groovy/life/qbic/portal/infrastructure/services/workflowtracking/WorkflowTrackingService.groovy:
```diff
@@ -15,21 +15,30 @@ class WorkflowTrackingService extends Service {
+        init()
+    }
+
+    WorkflowTrackingService(Service service) {
```
I suggest to get rid of this class. Mainly because it will produce confusion if we have several Service class extensions with different behaviour, which is not obvious from its name. 

Just create an `Endpoints` class, that holds the paths to the different endpoints as public static member variables, such as `WORKFLOWS`, `WORFLOWS_INFO`, etc, which you can then implement in your code with `Endpoints.WORKFLOWS`.

The complete URL shall remain dynamic, so you can keep doing: `new URL(service.rootUrl.toExternalForm() + Endpoints.WORKFLOWS)` later, when you make the call to the service.

KochTobi replied:
Will do that. 
The service class does not behave differently it only provides additional fields. One of the original reasons for implementing a helper class was that the Service class implements it's endpoints as field as well (health and routes endpoints). 
In addition to that I think that the ability to call `myservice.getWorkflowTracesEndpoint(runId)`
improves readability compared to 
`new URL(myservice.getRootUrl() + Endpoints.WORKFLOW_TRACES + runId)`

sven1103 commented on src/main/groovy/life/qbic/portal/infrastructure/services/workflowtracking/WorkflowTrackingServiceConnector.groovy:
```diff
@@ -1,8 +1,72 @@
+        }
+
+        for (service in serviceList) {
+            workflowTrackingServiceList.add(new WorkflowTrackingService((Service) service))
```
Exactly what I meant in my previous comment: this conversion of Service to WorkflowTrackingService is not intuitive.

KochTobi replied:
Yes this is bad, I agree :D

sven1103 commented on src/main/groovy/life/qbic/portal/infrastructure/services/workflowtracking/WorkflowTrackingServiceConnector.groovy:
```diff
@@ -1,8 +1,72 @@
+
+        def response = client.withCloseable { it.toBlocking().retrieve(request) }
+
+        return parser.parse(response.toCharArray())
```
This might fail, if you have an empty response?

KochTobi replied:
Retrieve returns a String. An empty response is an empty String in this case which results in a CharArray of size 0. I will make declare response to be of type String.

sven1103 replied:
It is a) unnecessary to make to conversion, as there is a method to parse String (http://docs.groovy-lang.org/2.4.0/html/gapi/groovy/json/JsonSlurper.html#parseText(java.lang.String)) and in any case: if the object is empty, for example when the query fails, you will run into an parse exception. So you have to check the HTTP response code.

KochTobi replied:
Ah I thought of the case that the response was empty but successful because there simply is no data.  Implemented Exception Handling and switched from receive to exchange.

sven1103 commented on src/main/groovy/life/qbic/portal/infrastructure/services/workflowtracking/WorkflowTrackingServiceConnector.groovy:
```diff
@@ -1,8 +1,72 @@
+    List<WorkflowTrackingService> workflowTrackingServiceList
 
+    WorkflowTrackingServiceConnector() {
+        this.serviceRegistryURL = loadServiceRegistryURLFromConfig()
```
A lot of stuff can go wrong here, and there is no exception handling.

Have a look how I solved it: https://github.com/qbicsoftware/sample-status-updater-cli/blob/master/sample-status-updater-application/src/main/java/life/qbic/samplestatusupdater/app/SampleUpdater.java

The Connector should be instantiated with a Service. It creates to much responsibility for the Connector class to also search for the correct service and handle exceptions. You will only create side effects that are difficult to test. 

You application code should search for a service and create the connector, otherwise it shall fail and notify the user.

KochTobi replied:
I did move the search for the service out of the application code by intent to keep the use case separated from the data source which can be basically anything as long as it implements the interface. I will however move the service search out of the connector class and handle it at initialisation in the `DependencyManager`. Would that suffice?

sven1103 replied:
with application code I meant the set up of all objects. Your entry point (main class) of your application. This is where you would handle exceptions that are related to initialization issues.

sven1103 reviewed (changes requested):
Some changes are required :)

KochTobi committed:
Replace WorkflowTrackingService class by static fields

rename WorkflowTrackingService to WorkflowTrackingServiceEndpoints
remove implementation of WorkflowTrackingService
add fields for the differen endpoints
modify listAllWorkflowRunInfo and listTracesForRunId in WorkflowTrackingServiceConnector
prepare separation of concerns between service search and service query

sven1103 commented on src/main/groovy/life/qbic/portal/infrastructure/services/workflowtracking/WorkflowTrackingServiceConnector.groovy:
```diff
@@ -16,11 +16,14 @@ import life.qbic.services.connectors.ConsulConnector
 
     URL serviceRegistryURL
-    List<WorkflowTrackingService> workflowTrackingServiceList
+    List<Service> workflowTrackingServiceList
```
No list required, a single object instance of type `Service` is sufficient :)

KochTobi replied:
Already implemented but not shown in the diff here

sven1103 commented on src/main/groovy/life/qbic/portal/infrastructure/services/workflowtracking/WorkflowTrackingServiceEndpoints.groovy:
```diff
@@ -0,0 +1,10 @@
+package life.qbic.portal.infrastructure.services.workflowtracking
+
+class WorkflowTrackingServiceEndpoints {
```
Enough to name it `Endpoints`, as the full class path gives you the context of the class: `life.qbic.portal.infrastructure.services.workflowtracking.Endpoints`

KochTobi committed:
Separate service request from service lookup

rename WorkflowTrackingServiceEndpoints to Endpoints
extract class for service lookup

KochTobi committed:
remove unneeded method

remove service registry loadin from config

sven1103 commented on src/main/groovy/life/qbic/portal/domain/listworkflowruns/WorkflowTrackingDataSource.groovy:
```diff
@@ -2,4 +2,8 @@ package life.qbic.portal.domain.listworkflowruns
 interface WorkflowTrackingDataSource {
 
+
+    def listAllWorkflowRunInfo()
```
```suggestion
    List<RunInfo> listAllWorkflowRunInfo()
```

sven1103 replied:
Interface methods should be declarative imo

sven1103 commented on src/main/groovy/life/qbic/portal/domain/listworkflowruns/WorkflowTrackingDataSource.groovy:
```diff
@@ -2,4 +2,8 @@ package life.qbic.portal.domain.listworkflowruns
 
+
+    def listAllWorkflowRunInfo()
+    def listTracesForRunId(String runId)
```
```suggestion
    List<Trace> listTracesForRunId(String runId)
```

KochTobi committed:
Update src/main/groovy/life/qbic/portal/domain/listworkflowruns/WorkflowTrackingDataSource.groovy

return type specification

Co-Authored-By: Sven F. <sven.fillinger@mailbox.org>

KochTobi committed:
Update src/main/groovy/life/qbic/portal/domain/listworkflowruns/WorkflowTrackingDataSource.groovy

return type specification

Co-Authored-By: Sven F. <sven.fillinger@mailbox.org>

KochTobi committed:
add missing imports

KochTobi committed:
type handling for datasource interface implementation

sven1103 commented on src/main/groovy/life/qbic/portal/infrastructure/services/ServiceFinder.groovy:
```diff
@@ -0,0 +1,33 @@
+            serviceList.addAll(factory.getServicesOfType(ServiceType.WORKFLOW_TRACKING))
+        }
+
+        if (serviceList.isEmpty()) {
```
This is boiler code. Just return the list. If it is empty, it is up to the implementation logic to decide what needs to be done. 

Like in the main class, print an error on the console if no service is found, otherwise continue with the application.

sven1103 replied:
Basically, this class is not needed, as I do not see its benefit for reusability.

KochTobi replied:
In this case this class has the purpose to separate the infrastructure and domain logic. The readability that is gained by having this in the `infrastructure.services` package justifies its existence in my opinion. Cramping everything in the messy Initialisation classes (DependencyManager and PortletEntryPointUI) would not fit in with the responsibilities of the respective classes.

KochTobi committed:
Replace exception throwing by logging a warning message.

remove error handling from ServiceFinder
add log.warn message for empty lists

KochTobi committed:
optimize imports

remove 1 unused import

KochTobi committed:
fix missing reference to serviceRegistryUrl

fix wrong ConsulConnector initialization

KochTobi committed:
Remove exception declaration

Remove `throws NullPointerException` from findWorkflowTrackingServices method

KochTobi committed:
Exception handling and switch from receive to exchange

add exception handling to queryServiceEndpoint method
refactor HttpClient method usage and switch from retrieve to exchange.

sven1103 reviewed (approved):
Looks good now!
